File: MicroBenchmarks/Ju/main_dot.jl

# initalize an array
function Init!(X,L)
    size=length(X)
    h=L/size
    for i=0:size-1
        if i>floor(size//8) && i<floor(size//2)+floor(size//8)
            X[i+1]=1.-2*(i-floor(size/8))*h/L;
        else
            X[i+1]=0.0
        end
    end
end



# proc[1-3]! functions make the same computation, with different programing
# styles.

function proc1!(X::Array{Float64,1},Y::Array{Float64,1},niter::Int)
 
    s=0.0
    for it=1:niter
        s=dot(X,Y)
        X,Y=Y,X
    end
    s
end
function proc2!(X::Array{Float64,1},Y::Array{Float64,1},niter::Int)
    const size=length(X)
    s=0.0
    for it=1:niter
        s=0.0
        @simd for i= 1:size
            s+=X[i]*Y[i]
        end
        X,Y=Y,X
    end
    s
end
function proc3!(X::Array{Float64,1},Y::Array{Float64,1},niter::Int)
    const size=length(X)
    s=0.0
    for it=1:niter
        s= sum(X.*Y)
        X,Y=Y,X
    end
    s
end
# measure as precisely as possible the execution time for a given
# p= proc[1-5]!
function test(p,X::Array{Float64,1},Y::Array{Float64,1},nit::Int)
   
    niter=nit
    Init!(X,1.)
    Init!(Y,1.)
 
    p(X,Y,niter)
    T=0.
    while true
        Init!(X,1.)
        Init!(Y,1.)
        t1 = time_ns()
        p(X,Y,niter)
        t = (time_ns() -t1)/niter
        if abs(t-T)/t<0.025
            break
    else
            T=t
            if niter> 150000
                error("too many iterations",niter)
            else
                niter*=2
            end
    end
    
end
    T,niter
end

# computation starts here:
size=16
sizemax=10^5
const niter=2
while size<sizemax
    println("size: ",size)
    X=Array{Float64}(size)
    Y=Array{Float64}(size)
    tbest=10.^29
    best=0
    t=0.0
    for p= [proc1!,proc2!,proc3!]
        t,it=test(p,X,Y,niter)
        if t<tbest
            tbest=t
            best=p
        end
        println(p," : t= ",t," nano-seconds ")
    end

    nflops=size*4
    flops=10^9*nflops/tbest
    println("\nbest: ",best)
    println("nb. flops (best): ",nflops, ", Gflops/s: ",flops/10^9)
    println("-------")
    
    size*=2
    println()
end

File: MicroBenchmarks/Py/main_dot.py
import numpy as np
import time

def Init(X,L):
    size=X.size
    h=L/size
    for i in range(0,size):
        if i>size//8 and i<size//2+size//8:
            X[i]=1.-2*(i-size//8)*h/L;
        else:
            X[i]=0.0


def dotd_1(X,Y,niter):
    size=X.size
    for it in range(0,niter):
        s=0.0
        for i in range(0,size):
            s+=X[i]*Y[i]
        X,Y=Y,X

def dotd_2(X,Y,niter):
    for it in range(0,niter):
        s= X.dot(Y)
        X,Y=Y,X  
      
def test(p,X,Y,nit):
   
    niter=nit
    T=0.
    while True:
        Init(X,1.)
        Init(Y,1.)
        t1 = time.time()
        p(X,Y,niter)
        treal=time.time() -t1
        t = treal/niter
        if treal>0.0001 and abs(t-T)/t<0.025:
            break
        else:
            T=t
            niter*=2

    return T,niter

size=16
sizemax=10**6
niter=10
parsef= lambda  f: str(f).split(" ")[1] #parse function name
while size<sizemax:
    print("size: ",size)
    X= np.empty(size)
    Y= np.empty(size)
    tbest=10.**20
    best=0
    t=0.0
    for p in  [dotd_1,dotd_2]:
        t,it=test(p,X,Y,niter)
        if t<tbest:
            tbest=t
            best=p
        print(parsef(p)," : t= ",t," seconds ")
    nflops=size*2
    flops=nflops/tbest
    print("\nbest: ",parsef(best))
    print("nb. flops (best): ",nflops, ", Gflops/s: ",flops/(10**9))
    print("-------")
    size*=2
    print(" ")
